### PR TITLE
[IMP] website: review `s_card` design and options

### PR DESCRIPTION
**Pull request by stefanorigano**
This PR significantly reworks the `s_card` component to improve its flexibility for use across various snippets. The primary focus is on enhancing customization options and ensuring seamless edition.

A major addition is the card "Cover Image" feature:
- Adjustable position layouts (top/left/right) with adaptive image's border radius.
- By default, the image size it's  based on container ratio, always covering the entire available space
- The newly introduced "Alignment" controller allows fine-tuning of the image focus-point , adjusting either vertically or horizontally based on the image ratio.
- Allow the removal of the current image OR, if not present, the injection of a new one.
  
 
| Cover at the Top - no ratio |Cover at the Top - use ratio | Cover one the side (left/right)  |
|--------|--------|--------|
| image | image | image|  

Color inheritance has been implemented, allowing the card to accept color-presets. This ensures improved readability for text, buttons, and links within the card.

| master | this PR |
|--------|--------|
| image | image | 

Additionally, users can now customize the card's border, shadow, gradients and shapes, providing greater control over its appearance.

| Example possible new designs |
|--------|
| image | 

To enhance layout consistency, particularly on mobile devices, a minimum width of `18rem` has been introduced. This value, quite arbitrary and chosen to match Bootstrap demo pages, ensures the card behaves as expected across different screen sizes. The implementation uses `min-width: MIN(18rem, 100%)` to maintain responsiveness without
exceeding the container's width.

| . | Card set to 25% (desktop) | Mobile |
|--------|--------|--------|
| 17.0 | image | image |
| this PR | image | image | 

task-3674888
part of task-3619705

Co-authored-by: @sobo-odoo 

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/scss/bootstrap_review_frontend.scss
Comment (sobo-odoo) [inline addons/web/static/src/scss/bootstrap_review_frontend.scss:15]: Here are the corrections I made (to make you gain some time after my long-awaited review :smiling_face_with_tear: )
Note that I am putting the whole files, as it is easier for me :wink: (you will see the diff when pasting the content in your editor :slightly_smiling_face:)

I added review comments on the commit so read them first (to understand what was wrong) and then you can use my code to fix the issues :slightly_smiling_face: (if you agree with my fixes of course) 
Note that I advise you to double check the behaviors (there is one that is weird but it was also the case before my changes (there is a "Beahviour" review comment about that) :+1: 

The fixes are mainly to:
- correctly use the options (so `computeWidgetVisibility`, `data-name`,...)
- removing the use of jQuery, 
- use double quotes in string,
- renaming some function into names I think are better (personal opinion, feel free to change them)
- + some website conventions

Comment (sobo-odoo) [reply addons/web/static/src/scss/bootstrap_review_frontend.scss:15]: `s_card.xml` file (note that I did not manage the version attributes (`data-vxml/vcss`):

```xml
<?xml version="1.0" encoding="utf-8"?>
<odoo>

<template id="s_card" name="Card">
    <div class="s_card o_card_img_top card o_cc o_cc1">
        <figure class="o_card_img_wrapper ratio ratio-16x9 mb-0">
            <img class="o_card_img card-img-top object-fit-cover" src="/web/image/website.s_card_default_image_1"/>
        </figure>
        <div class="card-body">
            <h5 class="card-title">Card title</h5>
            <p class="card-text">A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety of content, contextual background colors, and powerful display options.</p>
            <a t-att-href="cta_btn_href">Read More</a>
        </div>
    </div>
</template>

<template id="s_card_options" inherit_id="website.snippet_options">
    <xpath expr="." position="inside">
        <t t-call="website.snippet_options_background_options">
            <t t-set="selector" t-value="'.s_card'"/>
            <t t-set="with_colors" t-value="True"/>
            <t t-set="with_images" t-value="True"/>
            <t t-set="with_shapes" t-value="True"/>
            <t t-set="with_color_combinations" t-value="True"/>
            <t t-set="with_gradients" t-value="True"/>
        </t>

        <div data-js="Box" data-selector=".s_card">
            <t t-call="website.snippet_options_border_widgets"/>
            <t t-call="website.snippet_options_shadow_widgets"/>
        </div>

        <div data-js="CardWidth" data-selector=".s_card">
            <we-range string="Card Width"
                data-select-style="0"
                data-css-property="max-width"
                data-unit="%"
                data-min="8"
                data-max="100"
                data-display-range-value="true"
                data-display-range-value-unit="%"/>
        </div>

        <div data-js="CardImageOptions" data-selector=".s_card">
            <we-row string="Cover Image">
                <we-button title="Add" data-add-cover-image="" data-no-preview="true" class="o_we_bg_success" data-name="add_cover_image_opt">Add</we-button>
            </we-row>

            <we-row string="Cover Image">
                <we-button-group data-name="cover_image_position_opt">
                    <we-button title="Top"
                        data-name="card_img_top_opt"
                        data-select-class="o_card_img_top"
                        data-select-image-class="card-img-top"
                        data-img="/website/static/src/img/snippets_options/pos_top.svg"/>
                    <we-button title="Left"
                        data-select-class="o_card_img_horizontal flex-sm-row"
                        data-select-image-class="rounded-start"
                        data-img="/website/static/src/img/snippets_options/pos_left.svg"/>
                    <we-button title="Right"
                        data-select-class="o_card_img_horizontal flex-sm-row-reverse"
                        data-select-image-class="rounded-end"
                        data-img="/website/static/src/img/snippets_options/pos_right.svg"/>
                </we-button-group>
                <we-button title="Remove Cover" data-remove-cover-image="" data-no-preview="true" class="fa fa-trash-o fa-fw o_we_bg_danger" data-name="remove_cover_image_opt"/>
            </we-row>

            <we-select string="Ratio" data-apply-to=".o_card_img_wrapper" class="o_we_sublevel_1">
                <we-button data-select-class="">Image default</we-button>
                <we-button data-select-class="ratio ratio-1x1">Square</we-button>
                <we-button data-select-class="ratio ratio-4x3" data-dependencies="card_img_top_opt">Landscape - 4/3</we-button>
                <we-button data-select-class="ratio ratio-16x9" data-dependencies="card_img_top_opt">Wide - 16/9</we-button>
                <we-button data-select-class="ratio ratio-21x9" data-dependencies="card_img_top_opt">Ultrawide - 21/9</we-button>
                <we-button data-select-class="ratio o_card_img_ratio_custom" data-dependencies="card_img_top_opt" data-name="image_ratio_custom_opt">Custom</we-button>
            </we-select>

            <!-- Get bootstrap default column size -->
            <t t-set="colSize" t-value="100/12"/>
            <we-range string="Width" class="o_we_sublevel_1"
                title="Adjust the image width"
                data-name="cover_image_width_opt"
                data-dependencies="!card_img_top_opt"
                data-select-style="0"
                data-css-property="--card-img-size-h"
                data-unit="%"
                t-att-data-min="colSize"
                t-att-data-max="colSize * 11"
                t-att-data-step="colSize"/>

            <we-range
                string="Custom Ratio" class="o_we_sublevel_2"
                data-dependencies="image_ratio_custom_opt"
                data-select-style="0"
                data-css-property="--card-img-aspect-ratio"
                data-unit="%"
                data-display-range-value="true"
                data-display-range-value-unit="%"
                data-min="8"
                data-max="132"
                data-step="4"/>

            <we-range
                string="Alignment" class="o_we_sublevel_1"
                data-name="cover_image_alignment_opt"
                data-select-style="0"
                data-align-cover-image=""
                data-css-property="--card-img-ratio-align"
                data-unit="%"
                data-min="0"
                data-max="100"/>
        </div>
    </xpath>
</template>

<record id="website.s_card_000_scss" model="ir.asset">
    <field name="name">Card 000 SCSS</field>
    <field name="bundle">web.assets_frontend</field>
    <field name="path">website/static/src/snippets/s_card/000.scss</field>
</record>


</odoo>

```

File: addons/website/static/src/snippets/s_card/options.js
Comment (sobo-odoo) [inline addons/website/static/src/snippets/s_card/options.js:101]: Shouldn't the second one be `offsetWidth` ? :thinking: It doesn't work well with width though, maybe something to fix :slightly_smiling_face:

File: addons/web/static/src/scss/bootstrap_review_frontend.scss
Comment (sobo-odoo) [reply addons/web/static/src/scss/bootstrap_review_frontend.scss:15]: `options.js` file:
```js
/** @odoo-module **/

import { renderToElement } from "@web/core/utils/render";
import options from "@web_editor/js/editor/snippets.options";


options.registry.CardWidth = options.Class.extend({
    /**
     * @override
     */
    _computeWidgetState(methodName, params) {
        if (methodName === "selectStyle") {
            if (params.cssProperty === "max-width") {
                // If no `max-width` is set, consider it to be at 100%.
                if (!this.$target[0].style[params.cssProperty]) {
                    return "100%";
                }
            }
        }
        return this._super(...arguments);
    },
});

options.registry.CardImageOptions = options.Class.extend({
    /**
     * @override
     */
    start() {
        this._computeAlignmentDirection();
    },

    //--------------------------------------------------------------------------
    // Options
    //--------------------------------------------------------------------------

    /**
     * Injects a new cover image.
     */
    addCoverImage() {
        const imageWrapperEl = renderToElement("website.s_card.imageWrapper");
        this.$target[0].insertAdjacentElement("afterbegin", imageWrapperEl);
    },
    /**
     * Changes the cover image position.
     *
     * @see this.selectClass for parameters
     */
    selectImageClass(previewMode, widgetValue, params) {
        const imageEl = this.$target[0].querySelector(".o_card_img");
        for (const className of params.possibleValues) {
            if (className) {
                imageEl.classList.remove(className);
            }
        }
        imageEl.classList.add(widgetValue);
    },
    /**
     * Removes the cover image.
     */
    removeCoverImage() {
        const imageWrapperEl = this.$target[0].querySelector(".o_card_img_wrapper");
        imageWrapperEl.remove();
    },
    /**
     * Aligns the image inside the cover.
     */
    alignCoverImage() {
        this._computeAlignmentDirection();
    },

    //--------------------------------------------------------------------------
    // Private
    //--------------------------------------------------------------------------

    /**
     * @override
     */
    _computeWidgetVisibility(widgetName, params) {
        const hasCoverImage = !!this.$target[0].querySelector(".o_card_img_wrapper");
        if (widgetName === "add_cover_image_opt") {
            return !hasCoverImage;
        } else if (["cover_image_position_opt", "remove_cover_image_opt", "cover_image_width_opt"].includes(widgetName)) {
            return hasCoverImage;
        } else if (widgetName === "cover_image_alignment_opt") {
            const hasNonSquareRatio = !!this.$target[0].querySelector(".o_card_img_wrapper.ratio:not(.ratio-1x1)");
            return hasCoverImage && hasNonSquareRatio;
        }
        return this._super(...arguments);
    },
    /**
     * Changes the alignment widget behavior according to the image ratio.
     */
    _computeAlignmentDirection() {
        const imageEl = this.$target[0].querySelector(".o_card_img");
        const imageWrapperEl = this.$target[0].querySelector(".o_card_img_wrapper");
        if (!imageEl || !imageWrapperEl) {
            return;
        }

        const imgRatio = imageEl.offsetHeight / imageEl.offsetWidth;
        const wrapperRatio = imageWrapperEl.offsetHeight / imageWrapperEl.offsetHeight;
        const alignVertically = imgRatio <= wrapperRatio;
        this.$target[0].classList.toggle("o_card_img_adjust_v", alignVertically);
        this.$target[0].classList.toggle("o_card_img_adjust_h", !alignVertically);
    },
});

```

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
@stefanorigano  Here is my review :slightly_smiling_face:  This is a nice feature :+1: 

I have some comments and some questions :wink: There are also some behaviors that need be fixed (see review comments) :slightly_smiling_face:

Note that I fixed a lot of things while I was reviewing (in order to make you gain some time) => I added my fixes in my last review comments :wink: 
So read my review first, and then look at the fixes I made and use them (if you want) :slightly_smiling_face: 

Don't hesitate if you have questions (on some review comments or about the code itself) :slightly_smiling_face:

File: addons/website/static/src/snippets/s_card/options.js
Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/options.js:101]: Indeed, good catch! It was supposed to be `naturalWidth`, otherwise the size of the image it's not correctly computed because of object-fit 👍

File: addons/web/static/src/scss/bootstrap_review_frontend.scss
Comment (stefanorigano) [reply addons/web/static/src/scss/bootstrap_review_frontend.scss:15]: Unfortunately something was off compared to my initial implementation.
The custom ratio slider was visible for the horizontal layout as well. I addressed the issue in JS, let me know what you think!

---

Your poposal

https://github.com/user-attachments/assets/1340d946-8e30-4f1f-9789-a1210e41b72a

---

Old implementation 

https://github.com/user-attachments/assets/1bf4a22e-1f19-4f0c-81ac-45371856467d

File: addons/website/static/src/snippets/s_card/options.js
Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/options.js:12]: Not needed anymore indeed, thanks :)

File: addons/website/views/snippets/s_card.xml
Comment (stefanorigano) [reply addons/website/views/snippets/s_card.xml:98]: Makes slightly more sense indeed, thanks... but the visual result is a bit too noisy and the text would be truncated. I'm keeping `o_we_sublevel_1` 👍

File: addons/website/views/snippets/snippets.xml
Comment (stefanorigano) [reply addons/website/views/snippets/snippets.xml:467]: Good catch, thanks 👍

File: addons/web/static/src/scss/bootstrap_review_frontend.scss
Comment (sobo-odoo) [reply addons/web/static/src/scss/bootstrap_review_frontend.scss]: Yes I noticed it, it happens because of the "weird behavior" I mentioned [here](https://github.com/odoo/odoo/pull/172449#discussion_r1686927475) :wink:  
=> If the invalid ratio classes (so the ones that should not appear when the image is horizontal) are not removed when changing the option, the "Custom" value is still selected, and because of that, the "Custom Ratio" slider is still displayed (because of the dependencies) 
That's why I didn't do anything to hide it, because I first wanted to know if keeping "Custom" selected shouldn't be fixed first (if it is, the slider will automatically be hidden with it) :slightly_smiling_face:

**Comment by stefanorigano** [comment]
Hello @sobo-odoo , thanks so much for your detailed review and especially for speeding-up things by providing your code propositions directly 👍 
I've just pushed the necessary adaptations in a separate commit, let me know what you think, thanks!
Some notes:
-   I kept the "BS4 card-body override" in order to exclude migration issues. Also, on closer inspection, it still provides some useful adjustments for when then user choose to pick an arbitrary color rather than a preset. Arbitrary colors are not the best way to customise a card but, since the functionality it's there, these adaptation don't hurt 👍  
- Previously the "Alignment" range was hidden if the preferred ratio was set to "Square" :
```js
const hasNonSquareRatio = !!this.$target[0].querySelector(".o_card_img_wrapper.ratio:not(.ratio-1x1)");
```
This was conceptually wrong 'cause the fact that the desired result was a square didn't necessarily meant that the alignment option wasn't necessary anymore. The scenario we want to cover, instead, is that if the image and the wrapper share the same ratio, then the "Alignment" becomes unnecessary, thus can be hidden. In order to achieve this I create the `_getImageToWrapperRatio`.

- As discussed [here](https://github.com/odoo/odoo/pull/172449#discussion_r1687189818), there it was an issue when toggling the visibility of the "Custom Ratio" range slider. I do agree with your [analysis](https://github.com/odoo/odoo/pull/172449#discussion_r1687695333) . It would be better to recompute active options with an async function indeed, but I don't know how to address it 😕. The behaviour it's a "bit misleading" when switching between top and left/right indeed , but it's not critical IMO. I'll let you decide :)

Thanks

File: addons/website/views/snippets/s_card.xml
Comment (sobo-odoo) [reply addons/website/views/snippets/s_card.xml:88]: On a second thought, I think QSM does not like when we use `calc` :thinking: 
So instead what you can do is simply replace
```xml 
<t t-set="colSize" t-value="100/12"/>
```
by the bootstrap value 
```xml
<t t-set="colSize" t-value="8.33333333"/>
```
=> it will avoid having the rounding errors :slightly_smiling_face: 

And in the SCSS, replace the value of `--card-img-size-h` by `33.33333333%` (given what I wrote just above) :wink:

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
@stefanorigano Here is a second review :slightly_smiling_face: 

There are issues (with the backgrounds and some shadows) because of the BS3 rule and the `inherit` rules you added (surely to not consider the BS3 rules) :open_mouth: I added suggestions to fix them, see review comments :+1: 

I also tested more deeply than last time and noticed some behaviors that may need to be fixed :slightly_smiling_face: 

Note that I found a solution to force the "Ratio" option to "Square" when positioning the image horizontally (if there was an "invalid" ratio on it) => I added the code in my comments :wink: Note that depending on your answer on my "behavior" comment about the radius, the code may need to be adapted (don't hesitate to ask if you don't see how) :slightly_smiling_face:

File: addons/website/static/src/snippets/s_card/001.scss
Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/001.scss:18]: Thanks, great analysis 💯 

> either like I said with `.card:not([data-vcss]) .card-body`, or by finding a way to somehow overwrite it

I opted for `:where(.card:not([data-vcss])) .card-body` that has a [lower specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/:where).  
It solves any colors related issue for the new cards 👍 

---

> the background should only be applied on the card-body

It's not doable because Bootstrap cards could contain different type of content, like `.card-header`, `.card-footer`, `.list-group` and **multiple** `.card-body` 😅 ([ref](https://getbootstrap.com/docs/5.3/components/card/#kitchen-sink).)Here what you would get eventually  ⬇︎

image

...**but** there is no problem to solve IMO 🙂 To have **both** a cover image and a background image  it's a very corner case scenario... the only circumstance I can think of  it's when the background it's used a decoration, the same for shapes 😉  

image

File: addons/website/views/snippets/s_card.xml
Comment (stefanorigano) [reply addons/website/views/snippets/s_card.xml:7]: A bit debatable 'cause it really depend by what the user wants to achieve 🤔... but I do agree that `container` should cover the most plausible scenarios.. ok, let's do that 👍 

Be aware that it will invalidate the alignment option, I'll add a condition for this as well 😉

File: addons/website/views/snippets/s_card.xml
Comment (stefanorigano) [reply addons/website/views/snippets/s_card.xml:91]: Ok 🙂

File: addons/website/static/src/snippets/s_card/options.js
Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/options.js:82]: 🙄 indeed, the condition got lost during the last modifications, I'll re-introduce it 👍

**Comment by stefanorigano** [comment]
Thanks so much @sobo-odoo 🙌
Changes applied ([except applying colors to the body](https://github.com/odoo/odoo/pull/172449#discussion_r1690030867)), branch rebased and commits squashed 👍

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
@stefanorigano Here is a last review :slightly_smiling_face: 

There are review comments about (in chronological order) :
- last commit message nitpicks (we have to respect the title conventions :+1: )
-  the style fixes to make the background color work correctly 
- properly cleaning the DOM from unneeded classes and variables when removing the cover image (with code suggestions) (=> we don't like when the DOM is too polluted for nothing)
- other comments

Once this is done, I'll merge the PR :wink: (try to make the changes on only one push-force without any rebase in it so I can easily check the changes made :+1: you can rebase and push afterwards if you want :slightly_smiling_face:)

File: addons/website/static/src/snippets/s_card/001.scss
Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/001.scss:8]: I think that we cannot do that 🤔

- if people code  Bootstrap cards without enforcing a background-color, they would get a transparent card 
- If people customize the Bootstrap variable (eg. HTML/CSS Editor - > `user_custom_bootstrap_overridden.scss` -> `$card-bg: red;`) our changes will invalidate the customizations for no good reason 😕

image

This look like an editor general bug to me...  otherwise we have to assume that everything could be potentially customized should not have a default white background by default (unless I'm missing something)  🤷🏻‍♂️

Comment (sobo-odoo) [reply addons/website/static/src/snippets/s_card/001.scss:8]: For me this is not really an issue, any other snippet that has no color is transparent, so it should not be a problem with cards :thinking: In fact it will be consistent with them
Don't forget that this rule is exclusively for the new `s_card` snippet, and this snippet has a default color (the `o_cc1` you added in the template) so every dropped "Card" snippets will be o_cc1 by default :slightly_smiling_face: 
And the user can modify the color of the card by using the option (which should be the main way to change a card color) so there is no need for him to define the color by modifying the BS variable => he should set an inline style in the html editor in that case :thinking: 
I think it is more problematic to not be able to select a color with the option, than to not be able to set a custom variable (since there is the option for that) :slightly_smiling_face: 

When looking at the "Text-Highlight" inner snippet, which has the exact same background options, we see that it doesn't have the issue, because he doesn't have a variable-forced color => so this is really a card-only issue, which is why we think that ignoring the variable is the right solution :slightly_smiling_face: 

But if you are really not sure, we can ignore this for now and ask later if this is the right solution :wink:
Given that your PR is in master, we have time before people start using it :+1:  And the issue will not prevent your PR to be merged, since the templates are correct :wink: (I wanted to find what the issue really was first, because if it was a template issue, it would have impacted everything using the s_card later, but now I am sure the templates are correct :slightly_smiling_face:)

Comment (stefanorigano) [reply addons/website/static/src/snippets/s_card/001.scss:8]: > Don't forget that this rule is exclusively for the new s_card snippet

Right, default Bootstrap classes will not be affected indeed, this drastically reduce the issue indeed 👍 

> When looking at the "Text-Highlight" inner snippet, which has the exact same background options, we see that it doesn't have the issue, because he doesn't have a variable-forced color => so this is really a card-only issue,

Not sure because Bootstrap is wide spreading this approach, but we can address it another time indeed 👍 
I'll implement the change, thanks!

File: addons/web_editor/static/src/scss/web_editor.variables.scss
Comment (stefanorigano) [reply addons/web_editor/static/src/scss/web_editor.variables.scss]: This it **was** needed to address the scenario:
no preset -> white bg (bs default) -> white text (inherited from the main snippet container with a dak bg) 

Since ~cards~ `s_cards` "require" a preset (otherwise will be transparent), it not necessary indeed ;)
Thanks!

**Review by sobo-odoo — APPROVED** [review]
Alright, let's r+ ! :partying_face: 

@robodoo rebase-ff r+

**Pull request closed without merging**
